corehq/apps/consumer_user/models.py

```diff
@@ -0,0 +1,126 @@
+    """A user model that defines an individual patient, beneficiary, or consumer,
+    and allows this person to log in to a subset of HQ.
+
+    When authenticated, this user will have access to their data collected
+    about them by FLWs.
+
+    A ConsumerUser is not bound by a domain, and can access data collected
+    about them across the platform
+
+    This is linked with the regular Django User, as we use Django's auth to
+    limit access.
+
+    """
+    user = models.OneToOneField(User, on_delete=models.CASCADE)
+
+
+class ConsumerUserCaseRelationship(models.Model):
+    """An authorization model which links a ConsumerUser with a particular case
+    that they should have access to. A ConsumerUser can have many
+    ConsumerUserCaseRelationships.
+
+    ConsumerUsers can have access to the cases listed in this table, and any
+    subcases of that case.
+
+    """
+    consumer_user = models.ForeignKey(
+        ConsumerUser,
+        on_delete=models.CASCADE,
+        related_name='case_relationships',
+    )
+    case_id = models.CharField(max_length=255, unique=True)
+    domain = models.CharField(max_length=255)
+
+
+class ConsumerUserInvitation(models.Model):
+    """This model keeps track of the invitation status linked to a particular demographic case.
+
+    The invitation workflow is triggered when a case of type
+    CONSUMER_INVITATION_CASE_TYPE is created in the system. This case is
+    written back to as the stages of the invitation are carried out so an app
+    user can see what is happening, and can retrigger or change an invitation
+    if need be.
+
+    The invitation stores an email address - where the invite is sent, and the
+    demographic_case_id, which defines which case the user will have access to
+    once they accept the invite.
+
+    Once a ConsumerUserInvitation is accepted by a consumer, a new
+    ConsumerUserCaseRelationShip is created with the case_id set to the
+    invitation's demographic_case_id.
+
+    """
+    email = models.EmailField()
+    # The case id of the CONSUMER_INVITATION_CASE_TYPE which triggered this
+    # invitation to be sent
+    case_id = models.CharField(max_length=255)
+    # The case id of the highest ancestor case the user will have access to
+    # once they accept this invitation. This is normally a "demographic case",
+    # or the case which corresponds to the person.
+    demographic_case_id = models.CharField(max_length=255)
+    domain = models.CharField(max_length=255)
+    accepted = models.BooleanField(default=False)
+    invited_by = models.CharField(max_length=255)
+    invited_on = models.DateTimeField(auto_now_add=True)
+    active = models.BooleanField(default=True)
+
+    class Meta:
+        constraints = [
+            # There can be only one unaccepted invitation per demographic case
+            models.UniqueConstraint(
+                fields=['demographic_case_id'],
+                condition=models.Q(accepted=False, active=True),
```

Comment:
Hey @proteusvacuum 

Just checking if this should filter for just accepted?
Looks like it's not possible for an invitation to get inactive after being accepted, but if that is possible that can lead to two `ConsumerUserInvitation` for the same case?
Or there is a reason to filter for active as well?

Reply:
An invitation can be marked inactive regardless of whether it was accepted or not. We don't want to be able to create more than one active, un-accepted invitation at one time. 

In the case where there is already an accepted invitation for this demographic case, the following check is made, which writes back to the invitation case. That allows the user to understand what has happened.

https://github.com/dimagi/commcare-hq/pull/29348/files/387063095864ee79318f348491b5b78a0a2cff22#diff-4f70744434144cb96f67810cc84ff41973ef91613690c0de7031c867de7f7ea9R51

You could probably add a DB constraint for the scenario you are suggesting here, but I'm not sure it's worth it.

Reply:
When the invitation is accepted it isn't marked inactive, it is marked accepted:

https://github.com/dimagi/commcare-hq/blob/dd16681e0194ef43a7550bb05748d42d85582030/corehq/apps/consumer_user/models.py#L98-L101

Does this test help explain a bit further? https://github.com/dimagi/commcare-hq/blob/dbe928ce7033de211d156c2c935fc5db333a053c/corehq/apps/consumer_user/tests/test_signals.py#L113-L114

However it's totally possible that we missed a scenario when we wrote this initially. Do you think you could write a test to show what you think the right behaviour is, and whether that is currently handled by the code? 

The confusing part here is that there are two things called "invitation" - one is the case (`commcare-consumeruserinvitation`), and the other is the invitation object (`ConsumerUserInvitation`). In the scenario you just outlined, I don't think an invitation object (`ConsumerUserInvitation`) is ever created because of this: 

https://github.com/dimagi/commcare-hq/blob/e9829bff17437247bec912edf0854520ef87d6d4/corehq/apps/consumer_user/tasks.py#L49-L52

Reply:
Cool. I see, that's helpful, thanks @proteusvacuum